server/api/src/rsDataObjChksum.cpp

```diff
@@ -1,97 +1,400 @@
         free(fileStatOut);
         std::optional<rodsLong_t> size_in_database {};
-        auto select_and_condition { fmt::format("select DATA_SIZE where DATA_PATH = '{}'", dataObjInfo->filePath) };
-        for (auto&& row : irods::query{rsComm, select_and_condition}) {
+        auto select_and_condition { fmt::format("select DATA_SIZE where DATA_PATH = '{}'", dataObjInfo.filePath) };
+        for (auto&& row : irods::query{&rsComm, select_and_condition}) {
             try {
                 size_in_database = boost::lexical_cast<rodsLong_t>(row[0]);
-            } catch (const boost::bad_lexical_cast&) {
+            }
+            catch (const boost::bad_lexical_cast&) {
                 logger::api::error("_rsFileChksum: lexical_cast of [{}] for [{}] [{}] [{}] failed",
                     row[0],
-                    dataObjInfo->filePath,
-                    dataObjInfo->rescHier,
-                    dataObjInfo->objPath);
+                    dataObjInfo.filePath,
+                    dataObjInfo.rescHier,
+                    dataObjInfo.objPath);
                 return INVALID_LEXICAL_CAST;
             }
         }
+
         if (!size_in_database.has_value()) { return CAT_NO_ROWS_FOUND; }
         if (size_in_database.value() != size_in_vault) {
-            rodsLog(LOG_ERROR, "_rsFileChksum: file size mismatch. resource hierarchy [%s] vault path [%s] size [%ji] object path [%s] size [%ji]", dataObjInfo->rescHier, dataObjInfo->filePath,
-                    static_cast<intmax_t>(size_in_vault), dataObjInfo->objPath, static_cast<intmax_t>(size_in_database.value()));
+            rodsLog(LOG_ERROR, "_rsFileChksum: file size mismatch. resource hierarchy [%s] vault path [%s] size [%ji] object path [%s] size [%ji]",
+                    dataObjInfo.rescHier, dataObjInfo.filePath,
+                    static_cast<intmax_t>(size_in_vault), dataObjInfo.objPath, static_cast<intmax_t>(size_in_database.value()));
             return USER_FILE_SIZE_MISMATCH;
         }
+
         return 0;
+    } // verifyVaultSizeEqualsDatabaseSize
+
+    bool is_stale_replica(const DataObjInfo& _replica) noexcept
+    {
+        return _replica.replStatus == STALE_REPLICA;
+    } // is_stale_replica
+
+    bool is_member_of_bundle_resource(rodsLong_t _resource_id)
+    {
+        std::string resc_class;
+        const auto err = irods::get_resource_property<std::string>(_resource_id, irods::RESOURCE_CLASS, resc_class);
+        if (!err.ok()) {
+            logger::api::error(err.result());
+        }
+
+        return irods::RESOURCE_CLASS_BUNDLE == resc_class;
+    } // is_member_of_bundle_resource
+
+    std::vector<DataObjInfo*> get_good_replicas(DataObjInfo* _replicas)
+    {
+        std::vector<DataObjInfo*> good_replicas;
+        good_replicas.reserve(std::distance(begin(_replicas), end(_replicas)));
+
+        for (auto&& r : _replicas) {
+            if (!is_stale_replica(r) && !is_member_of_bundle_resource(r.rescId)) {
```

Comment:
careful here, opposite of stale is not necessarily good.    we now have intermediate...

Reply:
I agree. How should this believe when intermediate and lock states are encountered?

Reply:
*believe -> behave

Reply:
we should ignore anything other than 'good'.    right?

Reply:
i think you're right. with intermediates and logical locking, if someone opens a data object for write and another person tries to do checksum related operations, what do we expect to happen? my guess is that the checksum call would fail because the server said the data object is locked and being operated on. however, if that is not the case, then it may be possible for the checksum call to compute and update checksums for SOME of the replicas and not the ones being written to.